simulator/func_sim/syscall/mars_calls.h

```diff
@@ -0,0 +1,64 @@
+/*
+ * mars_calls.h - implementation of MARS syscalls
+ * @author Vyacheslav Kompan kompan.vo@phystech.edu
+ * Copyright 2018 MIPT-MIPS
+ */
+
+#ifndef MARS_CALLS_H
+#define MARS_CALLS_H
+
+/* Simulator modules */
+#include "func_sim/rf/rf.h"
+#include <infra/exception.h>
+#include "infra/types.h"
+/* Generic C & C++ */
+#include <cassert>
+#include <map>
+#include <memory>
+#include <iostream>
+
+
+struct TerminatedByExit final : Exception {
+    TerminatedByExit() : Exception( "Terminated", "exit called") {}
+};
+
+template <typename ISA>
+struct Syscall {
+    static std::unique_ptr<Syscall<ISA>> get_handler( bool ignore, RF<ISA> *rf, std::istream& cin = std::cin,
+                                                      std::ostream& cout = std::cout);
+    virtual void execute() {};
+
+    Syscall() = default;
+    virtual ~Syscall() = default;
+    Syscall( const Syscall&) = delete;
+    Syscall( Syscall&&) = delete;
+    Syscall& operator=( const Syscall&) = delete;
+    Syscall& operator=( Syscall&&) = delete;
+};
+
+
+template <typename ISA>
+class MARSCalls : public Syscall<ISA> {
+    using Register = typename ISA::Register;
+    using RegisterUInt = typename ISA::RegisterUInt;
+    using RegisterInt = sign_t<RegisterUInt>;
+
+    RF<ISA> *rf;
```

Comment:
What do you plan for memory?

Reply:
At first sight, we may keep a `FuncMemory` pointer in `MARSCalls`, as we do it in `FuncSim`, and set it with `FuncSim::set_memory()`.

Reply:
At the second sight, I think that exposing RF<ISA> is a bad idea, since it is a FuncSim internal, and it might not be used in PerfSim (e.g. if we implement register renaming).